### ISSUE TITLE
range types doesn't seem to be reset with implicit result variable in loop under `refc`

### ISSUE THREAD

**Issue by hamidb80** [nim-lang/Nim#22462]
### Description

```nim
type
    Object = object
        someComplexType: seq[int] # XXX works with `int` but not a complex type like `seq`
        index: Natural # XXX works with `int` but not `Positive`

func newObject(): Object =
    # XXX no problem with result = Object(index: 0)
    result.index.inc

for i in 1..3:
    let o = newObject()
    echo o.index
    # XXX no problem with echo newObject().index
```

### Nim Version

Nim Compiler Version 2.1.1 [Linux: amd64]
Compiled at 2023-08-11
Copyright (c) 2006-2023 by Andreas Rumpf

git hash: faf1c91e6a418e21d56ac6e45e1dbc851f9ffd22
active boot switches: -d:release

### Current Output

```text
1
2
3
```

### Expected Output
```text
1
1
1
```

### Additional Information

works with `--mm:arc` or `--mm:orc`

**Comment by juancarlospaco** [comment]
!nim c --gc:refc

```nim
type Object = object
  someComplexType: seq[int]
  index: Natural

func newObject(): Object = result.index.inc

for i in 1..3:
  let o = newObject()
  echo o.index
```

**Comment by hamidb80** [comment]
seems bug introduced in 1.4.0

**Comment by metagn** [comment]
!nim c --gc:refc

```nim
type Object = object
  someComplexType: seq[int]
  index: Natural
func newObject(): Object = result.index.inc
var s: seq[int]
for i in 1..3:
  let o = newObject()
  s.add o.index
doAssert s == @[1, 1, 1]
```

**Comment by ringabout** [comment]
fixed by #23215
